src/bindings/python/flux/cli/plugin.py

```diff
@@ -0,0 +1,187 @@
+##############################################################
+# Copyright 2024 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+##############################################################
+from os import getenv
+import glob
+from abc import ABC
+
+from flux.conf_builtin import conf_builtin_get
+from flux.importer import import_path
+
+
+class CLIPluginValue:
+    def __init__(self, value):
+        temp = value.split("=")
+        self.key = temp[0]
+        try:
+            self.value = temp[1]
+        except IndexError:
+            self.value = ""
+
+    def get_value(self):
+        """Getter for returning the key and value"""
+        return self.key, self.value
+
+
+class CLIPlugin(ABC):  # pragma no cover
+    """Base class for a CLI submission plugin
+
+    A plugin should derive from this class and implement one or more
+    base methods (described below)
+
+    Attributes:
+        opt (str): command-line key passed to -P/--plugin to activate the
+            extension
+        usage (str): short --help message for the extension
+        help (str): long --help message for the extension
+        version (str): optional version for the plugin, preferred in format
+            MAJOR.MINOR.PATCH
+        prog (str): command-line subcommand for which the plugin is active,
+            e.g. "submit", "run", "alloc", "batch", "bulksubmit"
+    """
+
+    def __init__(self, prog):
+        self.opt = None
+        self.usage = None
+        self.help = None
+        self.version = None
+        self.prog = prog
+        if prog.startswith("flux "):
+            self.prog = prog[5:]
+
+    def get_plugin_name(self):
+        """Return the KEY to invoke the plugin"""
+        return self.opt
+
+    def get_help_message(self):
+        """Return the message to print with --help"""
+        return self.help
+
+    def preinit(self, args, values):
+        """After parsing options, before jobspec is initialized
+
+        Either here, or in validation, plugin extensions should check types and
+        perform their own validation, since argparse will accept any string.
+
+        Args:
+            args (:py:obj:`Namespace`): Namespace result from
+                :py:meth:`Argparse.ArgumentParser.parse_args()`.
+            values (:obj:`Dict`): Dictionary of KEY[=VALUE] pairs returned
+                by parsing the arguments provided to the -P/--plugin
+                option. KEY is None if plugin was not invoked. VALUE is the
+                empty string if the plugin was invoked without VALUE.
```

Comment:
>  KEY is None if plugin was not invoked.

If `values` is a dictionary then by "KEY is None" here do you mean "is not set in the dictionary"?

Is `values` the dictionary of all `NAME=VALUE` pairs provided to `--plugin`? Maybe it would be better to use a `Namespace` here and keep the same semantics as argparse, i.e. `args.name` is `None` if the plugin was not invoked, `True` if it was invoked without an arg, and a string argument that the plugin is responsible for parsing if an argument was given.

Reply:
Oops, missed this on the last pass.